downloads/downloads-impl/src/main/java/com/duckduckgo/downloads/impl/DefaultFileDownloadNotificationManager.kt

```diff
@@ -21,69 +21,112 @@ import android.app.PendingIntent
-            val intent = Intent(Intent.ACTION_VIEW).apply {
-                setDataAndType(uri, mimeType)
-            }
+    @AnyThread
+    override fun showDownloadFinishedNotification(downloadId: Long, file: File, mimeType: String?) {
+        val filename = file.name
 
-            val pendingIntentFlags = PendingIntent.FLAG_IMMUTABLE or PendingIntent.FLAG_UPDATE_CURRENT
+        val intent = createIntentToOpenFile(applicationContext, file)
 
-            val notification = NotificationCompat.Builder(applicationContext, FileDownloadNotificationChannelType.FILE_DOWNLOADED.id)
-                .setContentTitle(filename)
-                .setContentText(applicationContext.getString(R.string.downloadComplete))
-                .setContentIntent(PendingIntent.getActivity(applicationContext, 0, intent, pendingIntentFlags))
-                .setAutoCancel(true)
-                .setSmallIcon(R.drawable.ic_file_download_white_24dp)
-                .build()
+        val pendingIntentFlags = PendingIntent.FLAG_IMMUTABLE or PendingIntent.FLAG_UPDATE_CURRENT
 
-            notificationManager.notify(NOTIFICATION_ID, notification)
-        }
+        val notification = NotificationCompat.Builder(applicationContext, FileDownloadNotificationChannelType.FILE_DOWNLOADED.id)
+            .setContentTitle(filename)
+            .setContentText(applicationContext.getString(R.string.downloadComplete))
+            .setContentIntent(PendingIntent.getActivity(applicationContext, 0, intent, pendingIntentFlags))
+            .setAutoCancel(true)
+            .setSmallIcon(R.drawable.ic_file_download_white_24dp)
+            .build()
+
+        notificationManager.cancel(downloadId.toInt())
+        // cancelling notifications is a deferred action, and I have seen how in subsequent notify calls end up in the wrong end state
+        waitForNotificationToBeCancelled(downloadId)
+        notificationManager.notify(downloadId.toInt(), notification)
     }
 
-    override fun showDownloadFailedNotification() {
-        mainThreadHandler().post {
-            val notification = NotificationCompat.Builder(applicationContext, FileDownloadNotificationChannelType.FILE_DOWNLOADED.id)
-                .setContentTitle(applicationContext.getString(R.string.downloadFailed))
-                .setSmallIcon(R.drawable.ic_file_download_white_24dp)
-                .build()
+    @AnyThread
+    override fun showDownloadFailedNotification(downloadId: Long, url: String?) {
+        val notification = NotificationCompat.Builder(applicationContext, FileDownloadNotificationChannelType.FILE_DOWNLOADED.id)
+            .setContentTitle(applicationContext.getString(R.string.downloadFailed))
+            .setSmallIcon(R.drawable.ic_file_download_white_24dp)
+            .apply {
+                url?.let { fileUrl ->
+                    val pendingIntent = PendingIntent.getBroadcast(
+                        applicationContext,
+                        1234,
```

Comment:
Use `downloadId.toInt()` instead of 1234.

Reply:
will do